### DISCUSSION TITLE
ページ内リンクについて

### DISCUSSION THREAD

**Discussion by blono** [increments/qiita-discussions#51] [要望・アイデア・不具合]
「# abc」
と見出しをつけた記事を作成すると、#abc のページ内リンクが作成されると思います。
例：https://qiita.com/{ユーザ名}/items/1234567#abc

このあと、記事作成者が見出しを
「# abcd」
とすると、https://qiita.com/{ユーザ名}/items/1234567#abc の URL は無効になり、代わりに https://qiita.com/{ユーザ名}/items/1234567#abcd が有効になると思います。

私見ですが、URL は互換性を維持しているほうが望ましいと思っているので、何か凄い何かをあれこれしてユーザが見出しの文字列を変更後も URL が維持されるような仕組みがあると嬉しいです！

**Comment by kskwtnk** [comment]
@blono 

ご意見ありがとうございます！
こちらについて、そもそも可能なのかなどが私では判断できず、1度社内で聞いてみます

その上でまたお返事しますね！

**Reply by kskwtnk** [reply]
@blono 

確認＆議論をしてきました！

## 結論

実現が不可能なわけではなさそうなものの、他の優先度の高い施策との兼ね合いですぐの対応は難しそうです 🙏 

## 議論の要約

- 過去に作成されたリンクを全て保存するのは出来なくはない
    - しかし相当数が多いのと、編集されたものを残すとハンドリングが難しそう
- アンカーを独自記法で設定できるようにする手もある
    - ただしプレーンなMarkdownから外れてしまうし、実装のコストもおそらく高め
- 方法自体はありそうなものの、工数の読めなさもあり、直近での対応を約束できない

**Reply by blono** [reply]
おお早速の議論ありがとうございます！
これ、確かに結構難しいんですよね、その割のリターンはどうなのか？というのに行き着くのは実は私も経験済みでした🙇‍♂️
markdown.js とか textile-js とか使うと簡単にその記方を達成できますが、カスタマイズして一旦そこから外れてしまう書き方を許容してしまうとバージョンアップに追従しにくかったりしますからね😓

私の場合、
\# {mylink}見出し

とすることで見出しは「見出し」としながらページ内リンクは \#mylink とするようにしました（「見出し」が「見出し2」になっても、URL の互換性を \#mylink で担保する）
単に「\# 見出し」となっているものについてはこれまで通りの動作（URL の互換性を担保しない、すべてのリンクを保存する方法はあまりに非現実的という判断）

いずれにせよ、優先順位がそれほど高くないのは間違いないです🙇

**Reply by oriom-dev** [reply]
全ての見出しにランダムなIDなどを割り振って、例えば見出しが「qiita」であれば「#qiita」「#eda3fea1」のどちらでもアクセスできるようにする。
記事内の内部リンクは、自動的に「#qiita」から「#eda3fea1」に置換して保存する。
見出しが「qoota」になってもIDは維持することで、URL の互換性を担保する。

この方法なら、過去に作成されたリンクを全て保存する必要はないので、比較的簡単に実装できるのではないでしょうか。

**Reply by kskwtnk** [reply]
@Lich496 

ご意見ありがとうございます！

実装の複雑さは減りそうですが、既存のIDの数の多さはあるので正直な話まだ工数や影響範囲が読めておりません……
ですが、また週明けに議論をしてみます 🙏

**Reply by munierujp** [reply]
GitHubでは、Markdownの見出しタグのかわりにHTMLの見出しタグを使い、そこにid属性をつけるという手法をとっていることがあります。

https://github.com/standard/standard/blob/master/docs/README-ja.md

Qiitaでもこの手法が使えるかと思います（Qiita自体のid属性と被らないような配慮は必要でしょうが）。

例：

```md
- [フー](#foo)

<h2 id="foo">フー</h2>
```

**Reply by kskwtnk** [reply]
@Lich496 @munierujp 

お返事が遅くなってしまいました 🙇 

## 結論

https://github.com/increments/qiita-discussions/discussions/51#discussioncomment-1505801 での内容とほぼ変わらず、実装できそうではあるものの今は取りかかれなさそう……となっています
たくさんご意見いただいているのに申し訳ないです 🙏 

---

https://github.com/increments/qiita-discussions/projects

現在進行中のプロジェクトである「エディタの改善」がかなり大きなものになっていまして、今回の同じタイミングで含めようと思うとかかる時間が膨らんでしまいそうでした

ご意見がたくさんあるので我々としても今回しっかり認知をしたのですが、こういった回答となってしまってすみません…… 🙇

**Comment by esperecyan** [comment]
QiitaにおいてはMarkdownの仕様変更や不具合により、見出しを変えずに記事を更新しただけでアンカーが変わってしまう、あるいは仕様変更や不具合により見出しを変更せざるを得なくなるという問題がこれまでに何度かありました。
別記事からのリンクが切れてしまう問題を避けるため、ぜひ改善をお願いしたいです。

**Reply by tanaka-asami415** [reply]
@esperecyan 
リクエストありがとうございます！
引き続き、他の優先度の高い施策との兼ね合いですぐの対応は難しい状況ですが
参考にさせていただきます🙏
